STYLE-GUIDE.md

```diff
@@ -0,0 +1,237 @@
+`.` and `$`, for example.  Use your best judgement here.
+
+### Do not use partial functions
+
+Never use a partial function when a safe version will work instead.  Do not use `error` or `undefined`.
+
+In some cases, partial functions are necessary, like indexing lists.  In this case, you must prove the safety of the function before using, ideally with an
+accompanying comment explaining the safety.
+
+### Prefer `map` to `fmap` or `<$>`
+
+When operating on a list, using `map` tells the reader "I'm transforming a list", where `fmap` tells them "I'm transforming *some* functor, and it doesn't matter which one.
+
+```haskell
+appendFoo t = t <> "-foo"
+txts = ["Hello", "darkness", "my", "old", "friend"]
+
+-- Ambiguous
+appendFoo <$> txts
+-- Slightly better
+fmap appendFoo txts
+-- Clear and Obvious
+map appendFoo txts
+```
+
+### Don't use `&` or `<&>`
+
+`&` and `<&>` are the flipped versions of `$` and `<$>`, and it is tempting to use them, since they read from left-to-right, like english.  However, the usefulness of
+`$` is to remove parentheses, and `&` makes that a lot harder:
+
+```haskell
+-- Raw parens
+foo (bar baz (quux))
+-- Using $
+foo $ bar baz $ quux
+-- Using &
+quux & bar baz & foo
+```
+
+`<$>` is meant to mimic `$`, but for functors.  Therefore, if we use `$` instead of `&`, we should also use `<$>` instead of `<&>`.
+
+## Data Types
+
+### Prefer `newtype`s to `type`s
+
+`newtype` values are zero-cost to construct and eliminate, and provide more informative error messages than `type` synonyms. Only use `type` for convenience aliases to existing types.
+
+### Don't use `String`
+
+`String` is almost always the wrong choice. If your type represents human-readable strings, use `Text`; if you have a blob of bytes, use `ByteString`. `-XOverloadedStrings` is enabled globally to make this easy.
+
+### Only use record selectors on single-constructor types
+
+The following code generates two partial functions, which is bad:
+
+``` haskell
+data Bad = Evil { getInt :: Int }
+         | Bad  { getFloat :: Float }
+
+-- getFloat (Evil 1) throws an error
+-- so does getInt (Bad 1)
+```
+
+Often, this is incorrect design.  You usually want something like the following three examples instead:
+
+```haskell
+-- Alternative 1
+data Good
+  = Okay Int
+  | NotBad Float
+
+-- Alternative 2
+data Alright = Alright
+  { getInt :: Maybe Int
+  , getFloat :: Maybe Float
+  }
+
+-- Alternative 3
+-- This is a little bit heavyweight, but is really strongly typed.
+newtype SpecialInt = SpecialInt { getInt :: Int }
+newtype SpecialFloat = SpecialFloat { getFloat :: Float }
+data Great
+  = NiceInt SpecialInt
+  | CoolFloat SpecialFloat
+```
+
+## Imports
+
+### Don't qualify imports with single letters
+
+Qualified imports deserve intelligent names, like any other variable.  Some common instances are caught by `hlint`:
+
+```haskell
+import Data.Text qualified as T -- bad; hlint will catch this specific case
```

Comment:
I thought we said we were fine with certain imports like Text and ByteString which have widely accepted short qualified imports being left alone? `T`, `BS`, and some others. I'm ok with changing what we said, I just want to make sure that we're no longer ok with short imports for `Text` or `Bytestring`.

Reply:
I was told that we decided on names rather than letters.  It was decided during a retro that I was not present for, and we have already begun checking it with `hlint`

Reply:
Ahhh, gotcha, yea we decided to use full words except for a few common exceptions like `ByteString`. @cnr do you have any input on this?

Reply:
exceptions might be nice for:

- `ByteString` - `BS`
- lazy `ByteString` - `BL`

we don't really use lazy `Text` anywhere -- otherwise I'd want `T` and `TL` for those